[Program Capture] higher order primitives support pytree input and output (#6081)
PennyLaneAI/pennylane: pennylane/capture/explanations.md

a constant.  The arguments to the `repeat` primitive are `n (const a) x (hardcoded 2.0=y)`.
You can also see the const variable `a` as argument `e:i32[]` to the inner nested jaxpr as well.

### Pytree handling

> Evaluating a jaxpr requires a flat list of tensor-like inputs and returning a flat

Reviewer: ```suggestion
Evaluating a jaxpr requires accepting and returning a flat list of tensor-like inputs and outputs.
```

> list of tensor-like outputs.  But these long lists can be hard to manage and are very
> restrictive on the allowed functions. But we can take advantage of pytrees to allow handling

Reviewer 2: ```suggestion
list of tensor-like outputs.  These long lists can be hard to manage and are very
restrictive on the allowed functions, but we can take advantage of pytrees to allow handling
```
Very minor copy suggestion :+1:

arbitrary functions.

> To start, we import the `FlatFn` helper. This class converts a function to one caches

Reviewer 2: ```suggestion
To start, we import the `FlatFn` helper. This class converts a function to one that caches
```

the resulting result pytree into `flat_fn.out_tree` when executed. This can be used to repack the
results into the correct shape. It also returns flattened results. This does not particularly
matter for program capture, as we will only be producing jaxpr from the function, not calling
it directly.

The following demonstrates how the tree utilities and the `FlatFn` class
can be used together to unpack and repack the variables:

```pycon
>>> from pennylane.capture.flatfn import FlatFn
>>> def f(x): # define a function with pytree inputs and outputs
...     return {"a": x[0], "b": x[1]+1}
>>> args = ([0.1, 0.2],) # the arguments to the function
>>> flatfn = FlatFn(f)
>>> flatfn.out_tree is None # initialized to None
True
>>> results = flatfn(*args)
>>> results
[0.1, 1.2]
>>> flatfn.out_tree # set once function is called
PyTreeDef({'a': *, 'b': *})
>>> jax.tree_util.tree_unflatten(flatfn.out_tree, results)
{'a': 0.1, 'b': 1.2}
```

> Using these tools, we can now redefine our wrapper around the repeat primitive. This
> code just extends what we did with the repeat primitive above.

Reviewer 2: ```suggestion
Using these tools, we can now redefine our wrapper around the `repeat` primitive. This
code just extends what we did with the `repeat` primitive above.
```

a constant.  The arguments to the `repeat` primitive are `n (const a) x (hardcoded 2.0=y)`.
You can also see the const variable `a` as argument `e:i32[]` to the inner nested jaxpr as well.

### Pytree handling

> Evaluating a jaxpr requires a flat list of tensor-like inputs, and returns a flat

Reviewer: ```suggestion
Evaluating a jaxpr requires a flat list of tensor-like inputs and returning a flat
```

list of tensor-like outputs.  But these long lists can be hard to manage and are very
restrictive on the allowed functions. But we can take advantage of pytrees to allow handling
arbitrary functions.

To start, we import the `FlatFn` helper. This class converts a function to one that

> accepts flat inputs and returns flat outputs.  It also caches the result pytree

Reviewer 3: This would have to be updated as well with the acceptance of unflattened args.

definition into the `flat_fn.out_tree` property. This can be used to repack the
results into the correct shape.

The following demonstrates how the tree utilities and the `FlatFn` class
can be used together to unpack and repack the variables:

```pycon
>>> from pennylane.capture.flatfn import FlatFn
>>> def f(x): # define a function with pytree inputs and outputs
...     return {"a": x[0], "b": x[1]+1}
>>> args = ([0.1, 0.2],) # the arguments to the function
>>> flat_args, in_tree = jax.tree_util.tree_flatten(args)
>>> flat_args
[0.1, 0.2]
>>> in_tree
PyTreeDef(([*, *],))
>>> flatfn = FlatFn(f)
>>> flatfn.out_tree is None # initialized to None
True
>>> results = flatfn(*args)
>>> results
[0.1, 1.2]
>>> flatfn.out_tree # set once function is called
PyTreeDef({'a': *, 'b': *})
>>> jax.tree_util.tree_unflatten(flatfn.out_tree, results)
{'a': 0.1, 'b': 1.2}
```

> Using these tools, we can now redefine out wrapper around the repeat primitive.

Reviewer 3: ```suggestion
Using these tools, we can now redefine our wrapper around the repeat primitive.
```

a constant.  The arguments to the `repeat` primitive are `n (const a) x (hardcoded 2.0=y)`.
You can also see the const variable `a` as argument `e:i32[]` to the inner nested jaxpr as well.

### Pytree handling

> Evaluating a jaxpr requires a flat list of tensorlike inputs, and returns a flat
> list of tensorlike outputs.  But these long lists can be hard to manage and are very
> restrictive on the allowed functions. But we can take advantage of pytree to allow handling
> arbitrary functions.

Reviewer 3: ```suggestion
Evaluating a jaxpr requires a flat list of tensor-like inputs, and returns a flat
list of tensor-like outputs.  But these long lists can be hard to manage and are very
restrictive on the allowed functions. But we can take advantage of pytrees to allow handling
arbitrary functions.
```

To start, we import the `FlatFn` helper. This class converts a function to one that
accepts flat inputs and returns flat outputs.  It also caches the result pytree
definition into the `flat_fn.out_tree` property. This can be used to repack the
results into the correct shape.

The following demonstrates how the tree utilities and the `FlatFn` class
can be used together to unpack and repack the variables:

```pycon

> >>> from pennylaane.capture.flatfn import FlatFn

Author: ```suggestion
>>> from pennylane.capture.flatfn import FlatFn
```

>>> def f(x): # define a function with pytree inputs and outputs
...     return {"a": x[0], "b": x[1]+1}
>>> args = ([0.1, 0.2],) # the arguments to the function
>>> flat_args, in_tree = jax.tree_util.tree_flatten(args)
>>> flat_args
[0.1, 0.2]
>>> in_tree
PyTreeDef(([*, *],))

> >>> flatfn = FlatFn(f)

Author: ```suggestion
>>> flatfn = FlatFn(f, in_tree)
```

>>> flatfn.out_tree is None # initialized to None
True
>>> results = flatfn(*flat_args)
>>> results
[0.1, 1.2]
>>> flatfn.out_tree # set once function is called
PyTreeDef({'a': *, 'b': *})
>>> jax.tree_util.tree_unflatten(flatfn.out_tree, results)
{'a': 0.1, 'b': 1.2}
```

Using these tools, we can now redefine out wrapper around the repeat primitive.

```python

> def repeat(funcm, n: int):

Reviewer 3: ```suggestion
def repeat(func, n: int):
```

    def new_func(*args, **kwargs):

        func_bound_kwargs = partial(func, **kwargs)

        flat_args, in_tree = jax.tree_util.tree_flatten(args)
        flat_fn = FlatFn(func_bound_kwargs, in_tree)

        jaxpr = jax.make_jaxpr(flat_fn)(*flat_args)
        n_consts = len(jaxpr.consts)

>         results = repeat_prim.bind(n, *jaxpr.consts, *flat_args, jaxpr=jaxpr.jaxpr, n_consts=n_consts)

Reviewer 3: What's `repeat_prim`?

Author: Defined in the previous section.
